website/www/site/content/en/documentation/programming-guide.md

```diff
@@ -5188,16 +5188,63 @@ restriction pairs.
+3. [ByteKeyRangeTracker](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.html)
+
+The SDF also has a built-in `RestrictionTracker` implementation in Python:
+1. [OffsetRangeTracker](https://beam.apache.org/releases/pydoc/current/apache_beam.io.restriction_trackers.html#apache_beam.io.restriction_trackers.OffsetRestrictionTracker)
+
+The watermark state is a user-defined object which is used to create a `WatermarkEstimator` from a
+`WatermarkEstimatorProvider`. The simplest watermark state could be a `timestamp`.
+
+The watermark estimator provider lets SDF authors define how to initialize the watermark state and
+create a watermark estimator. In [Java](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/DoFn.ProcessElement.html)
+this is the `DoFn`. [Python](https://beam.apache.org/releases/pydoc/current/apache_beam.transforms.core.html#apache_beam.transforms.core.WatermarkEstimatorProvider)
+has a dedicated `WatermarkEstimatorProvider` type.
+
+The watermark estimator tracks the watermark when an element-restriction pair is in progress.
+For APIs details, read the [Java](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/splittabledofn/WatermarkEstimator.html)
+and [Python](https://beam.apache.org/releases/pydoc/current/apache_beam.io.iobase.html#apache_beam.io.iobase.WatermarkEstimator)
+reference documentation.
+
+There are some built-in `WatermarkEstimator` implementations in Java:
+1. [Manual](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/splittabledofn/WatermarkEstimators.Manual.html)
+2. [MonotonicallyIncreasing](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/splittabledofn/WatermarkEstimators.MonotonicallyIncreasing.html)
+3. [WallTime](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/splittabledofn/WatermarkEstimators.WallTime.html)
+
+Along with the default `WatermarkEstimatorProvider`, there are the same set of built-in
+`WatermarkEstimator` implementations in Python:
+1. [ManualWatermarkEstimator](https://beam.apache.org/releases/pydoc/current/apache_beam.io.watermark_estimators.html#apache_beam.io.watermark_estimators.ManualWatermarkEstimator)
+2. [MonotonicWatermarkEstimator](https://beam.apache.org/releases/pydoc/current/apache_beam.io.watermark_estimators.html#apache_beam.io.watermark_estimators.MonotonicWatermarkEstimator)
+3. [WalltimeWatermarkEstimator](https://beam.apache.org/releases/pydoc/current/apache_beam.io.watermark_estimators.html#apache_beam.io.watermark_estimators.WalltimeWatermarkEstimator)
 
 To define an SDF, you must choose whether the SDF is bounded (default) or
 unbounded and define a way to initialize an initial restriction for an element.
```

Comment:
Perhaps you can explain how does one know if it's bounded / unbounded? 

I took a stab at it here, but feel free to write your own if you think this is not good: 
```
Bounded DoFns are those where the work represented by an element is well known a priori, and that 
it has an end. Unbounded are those where the amount of work does not have a specific end, or the 
amount of work is not known a priori. The boundedness/unboundedness of your SDF has implications
 for Bundle Finalization(link).
```

Perhaps it's useful to mention examples?
```
An example of an unbounded element would be a Kafka or a PubSub topic; while a bounded element 
may represent a file, or group of files.
```

Reply:
WDYT about this? 

The boundedness and unboundedness of your SDF has implications for Bundle Finalization(link). The distinction is based on how the amount of work is represented: 

* Bounded DoFns are those where the work represented by an element is well-known beforehand and has an end. Examples of bounded elements include a file or group of files.
* Unbounded DoFns are those where the amount of work does not have a specific end or the 
amount of work is not known befrehand.  Examples of unbounded elements include a Kafka or a PubSub topic.

Reply:
Updated the guide with suggestion. I removed `The boundedness and unboundedness of your SDF has implications for Bundle Finalization(link).` since boundedness/unboundedness is not related to bundle finalization nor drain. When draining, it will look into the `isBounded` API. They are separated concept. Thanks for your help!